### ISSUE TITLE
Running command that pipes to another command

### ISSUE THREAD

**Issue by lordcodes** [lordcodes/turtle#104]
**Describe the problem**

Currently, if run a command that pipes to another command, it doesn't pipe the output from the first command into the second command as it would if ran on the command line.

For example:
`echo '{"key":"value"}' | jq '.'`

**Describe your solution**

The commands are executed correctly, if the output of the first command being piped into the second command.

**Checklist**

- [X] I've read the [guide for contributing](https://github.com/lordcodes/turtle/blob/master/CONTRIBUTING.md).
- [X] I've checked there are no other [open pull requests](https://github.com/lordcodes/turtle/pulls) for the feature.
- [X] I've checked there are no other [open issues](https://github.com/lordcodes/turtle/issues) for the same feature.

**Additional context**

It looks like a change may be made with how the command(s) are prepared and built using `ProcessBuilder` so that they are executed correctly.

**Workaround**
Currently, you can run multiple commands by simply capturing the output of the first command into a Kotlin property and then running the second command using its value as input, rather than using pipe within the 'shell' command.

**Comment by jmfayard** [comment]
Maybe it's just me, but I prefer Kotlin typesafe syntax to Bash messy [primitive obsession](https://refactoring.guru/smells/primitive-obsession) (everything is a string).

So I was thinking about a more functional approach where you create Commands and Pipelines,  
and only at the last step you execute it

```kotlin
/**
 * Pure function that represents but doesn't execute this command:
 *  $ find . -type f | grep '.kt' | xargs grep 'fun ' 2> /dev/null | wc -l
 */
fun measureKotlinFun(): Pipeline {
    val findFiles = Command("find", ".", "-type", "f")
    val onlyKotlinFiles = Command("grep", ".kt")
    val foreachSearchFun = Command("xargs", "grep", "fun ")
        .withRedirect(CommandRedirect(stderrIgnore = true))
    val countLines = Command("wc", "-l")
    val pipeline = findFiles pipe onlyKotlinFiles pipe foreachSearchFun pipe countLines
    return pipeline
}

fun imperativeShell() {
    measureKotlinFun()
         .withRedirect(CommandRedirect(stdout = output))
        .executeCommand()
}
```

https://github.com/jmfayard/turtle/blob/58c1544282d2a76ab3cdf774cab5bc99517489d0/turtle/src/main/kotlin/com/lordcodes/turtle/Command.kt#L25-L37

WDT?

**Comment by lordcodes** [comment]
Interesting idea. Yeah I actually did look into supporting 'pipe' a bit and didn't really get anywhere. I think it's a little complicated due to how the commands are ran using Process and how it captures standard out and in. Based on that I was thinking of just closing this issue and to say to just run commands one at a time, capturing output of one and passing into the next one using Kotlin properties.

I like the idea of what you have shown here with the pipeline. The issue so far is I have not managed to actually get piped commands to work, i.e. for a command to take output of previous command and pipe it in automatically.

Currently, I just do one command after another and use output of a previous command as an argument to a later command.

**Comment by jmfayard** [comment]
> The main blocker is working out how to implement the actual piping

I got it to work using https://github.com/zeroturnaround/zt-exec which seems like the right library to implement my Pipeline class

Proof of concept here: https://github.com/jmfayard/turtle/commit/51aff7ddb76102a4c6dad43ae1f5a5333ed6939f

**Comment by lordcodes** [comment]
Nice concept. We should be able to take the plumbing required to forward output of one to input of next without the dependency. Looks like a great starting point.

**Comment by jmfayard** [comment]
Yes but you would have to reinvent all the threading etc... yourself. 
Personally I would start with the dependency as an implementation detail, if that works well you can always reinvent the wheel and remove it later.

**Comment by lordcodes** [comment]
Threading? I thought it was just reading from the input stream.

I understand. The issue is that Turtle was written as a simple library to run commands. This dependency is essentially a Java version of what Turtle is.

**Comment by jmfayard** [comment]
Maybe I'm missing something, but to create a pipeline, don't you have to launch each program in a separate thread?
How would blocking calls work without that?

**Comment by lordcodes** [comment]
Maybe I just don't see it until I would try and work on it, however, there is no threading now and the pipeline would be running the commands sequentially. I don't see the difference to just running one command at a time in a process like they are now, but that each commands reads in from stream.

Anyway, realistically I'm not going to look into this or work on this anytime soon. If you would like this soon, then you are welcome to work on something. However, I'm really not a fan of introducing that library as a dependency, due to the fact it is a Java library with no recent changes. If threading and a lot of complexity is required to implement this, i.e. it isn't as simple as it seems to me 🤣, pipelines could be built as a separate library potentially.

**Comment by jmfayard** [comment]
> Maybe I just don't see it until I would try and work on it, however, there is no threading now and the pipeline would be running the commands sequentially. I don't see the difference to just running one command at a time in a process like they are now, but that each commands reads in from stream.

There is a big different with

>  yes "Say yes indefinitely" | head -5

and 

> yes "Say yes indefinitely" > /tmp/yes.txt
> head -5 < /tmp/yes.txt

The first pipeline ends in a few miliseconds and prints five times "Say yes indefinitely"
The second is an infinite loop.

Here my version doesn't do an infinite loop

```kotlin
val inifiniteCommand = Command("yes", "say yes indefinitely")
val pipeline2 = inifiniteCommand `|` Command("head", "-5")
pipeline2.execcute()
```

https://github.com/jmfayard/turtle/commit/51aff7ddb76102a4c6dad43ae1f5a5333ed6939f?diff=split

**Comment by jmfayard** [comment]
If the dependency is an issue, we can put the pipeline feature in a
separate module.

Also it was fun to discuss but I don't really need pipelines right now.

Having the Command abstraction on the other hand would be really helpful.



Le jeu. 11 août 2022 à 13:26, Andrew Lord ***@***.***> a
écrit :

**Comment by lordcodes** [comment]
I agree. We can create a separate Issue to track that.

**Comment by parrotcar00** [comment]
I came across this project: https://github.com/dsherret/dax which is deno/typescript library for shell command running.

I think turtle is very similar in functionality to the above.
I was looking to see how to implement the shell pipe operator in turtle but this issue tells me it's on the todo list.

The dax project above implements the pipe operator like:
```
const output = await $`echo foo && echo bar`
  .pipe($`grep foo`)
.text();
```
Maybe this can be of some help while implementing the pipe feature?
